frame/nfts/src/features/atomic_swap.rs

```diff
@@ -23,53 +23,72 @@ use frame_support::{
 
 impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	pub fn do_create_swap(
-		sender: T::AccountId,
-		collection_id: T::CollectionId,
-		item_id: T::ItemId,
+		caller: T::AccountId,
+		offered_collection_id: T::CollectionId,
+		offered_item_id: T::ItemId,
 		desired_collection_id: T::CollectionId,
-		desired_item_id: T::ItemId,
+		maybe_desired_item_id: Option<T::ItemId>,
 		maybe_price: Option<ItemPrice<T, I>>,
+		maybe_price_direction: Option<PriceDirection>,
 		maybe_duration: Option<<T as SystemConfig>::BlockNumber>,
 	) -> DispatchResult {
-		let item = Item::<T, I>::get(&collection_id, &item_id).ok_or(Error::<T, I>::UnknownItem)?;
-		ensure!(item.owner == sender, Error::<T, I>::NoPermission);
-
-		Item::<T, I>::get(&desired_collection_id, &desired_item_id)
+		let item = Item::<T, I>::get(&offered_collection_id, &offered_item_id)
 			.ok_or(Error::<T, I>::UnknownItem)?;
+		ensure!(item.owner == caller, Error::<T, I>::NoPermission);
+
+		match maybe_desired_item_id {
+			Some(desired_item_id) => {
+				if !(Item::<T, I>::contains_key(&desired_collection_id, &desired_item_id)) {
+					return Err(Error::<T, I>::UnknownItem.into())
+				}
+			},
+			None =>
+				if !(Collection::<T, I>::contains_key(&desired_collection_id)) {
+					return Err(Error::<T, I>::UnknownCollection.into())
+				},
+		};
+
+		if maybe_price.is_some() && maybe_price_direction.is_none() {
+			return Err(Error::<T, I>::PriceDirectionNotSet.into())
+		}
+		if maybe_price.is_none() && maybe_price_direction.is_some() {
+			return Err(Error::<T, I>::PriceDirectionCannotBeSet.into())
+		}
```

Comment:
is a match a bit clearer?
```suggestion
       match (maybe_price, maybe_price_direction) {
           (Some(_), None) => Err(Error::<T, I>::PriceDirectionNotSet.into()),
           (None, Some(_)) => Err(Error::<T, I>::PriceDirectionCannotBeSet.into()),  
		   _ => {}
		}
```

Reply:
replaced with `PriceWithDirection` struct